app/src/main/java/fi/aalto/legroup/achso/storage/local/OptimizedLocalVideoRepository.java

```diff
@@ -47,62 +44,75 @@ public FindResults getAll() throws IOException {
             throw new IOException("Couldn't list files in " + storageDirectory);
         }
 
-        ArrayList<FindResult> results = new ArrayList<>(manifests.length);
+        videos = new HashMap<>(manifests.length * 2);
+        allResults = new ArrayList<>(manifests.length);
         for (File manifest : manifests) {
-            results.add(new FindResult(getIdFromManifest(manifest), manifest.lastModified()));
-        }
 
-        return new FindResults(results);
-    }
-
-    @Override
-    public long getLastModifiedTime(UUID id) throws IOException {
+            Video video = serializer.load(Video.class, manifest.toURI());
 
-        File manifest = getManifestFromId(id);
-        return manifest.lastModified();
-    }
+            video.setManifestUri(Uri.fromFile(manifest));
+            video.setRepository(this);
+            video.setLastModified(new Date(manifest.lastModified()));
 
-    @Override
-    public VideoInfo getVideoInfo(UUID id) throws IOException {
+            OptimizedVideo optimized = new OptimizedVideo(video);
+            videos.put(optimized.getId(), optimized);
 
-        File manifest = getManifestFromId(id);
-        VideoInfo videoInfo = serializer.load(VideoInfo.class, manifest.toURI());
+            allResults.add(new FindResult(getIdFromManifest(manifest), manifest.lastModified()));
+        }
+    }
 
-        videoInfo.setManifestUri(Uri.fromFile(manifest));
+    protected File getManifestFromId(UUID id) {
+        return new File(storageDirectory, id.toString() + ".json");
+    }
 
-        return videoInfo;
+    private UUID getIdFromManifest(File manifest) {
+        String basename = Files.getNameWithoutExtension(manifest.getName());
+        return UUID.fromString(basename);
     }
 
     @Override
-    public Video getVideo(UUID id) throws IOException {
-
-        File manifest = getManifestFromId(id);
-        Video video = serializer.load(Video.class, manifest.toURI());
-
-        video.setManifestUri(Uri.fromFile(manifest));
-        video.setRepository(this);
+    public FindResults getAll() throws IOException {
+        return new FindResults(allResults);
+    }
 
-        return video;
+    @Override
+    public OptimizedVideo getVideo(UUID id) {
+        return videos.get(id);
     }
 
     @Override
     public void save(Video video) throws IOException {
+        videos.put(video.getId(), new OptimizedVideo(video));
+
         serializer.save(video, getManifestFromId(video.getId()).toURI());
+
+        // FIXME: This is stupid
+        refresh();
         bus.post(new VideoRepositoryUpdatedEvent(this));
     }
 
     @Override
+    public void save(OptimizedVideo video) throws IOException {
+        save(video.inflate(savePool));
+        savePool.release();
+    }
+
+    @Override
     public void delete(UUID id) throws IOException {
         File manifest = getManifestFromId(id);
 
         if (manifest.delete()) {
+
+            // FIXME: This is stupid
```

Comment:
Those will be fixed when there will be more repositories and I get some view of the bigger picture, now it's too early to decide still (for me)